core/trino-parser/src/main/java/io/trino/sql/jsonpath/PathParser.java

```diff
@@ -0,0 +1,165 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.jsonpath;
+
+import io.trino.jsonpath.JsonPathBaseListener;
+import io.trino.jsonpath.JsonPathLexer;
+import io.trino.jsonpath.JsonPathParser;
+import io.trino.sql.jsonpath.tree.PathNode;
+import io.trino.sql.parser.ParsingException;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonToken;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.atn.PredictionMode;
+import org.antlr.v4.runtime.misc.Pair;
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public final class PathParser
+{
+    private final BaseErrorListener errorListener;
+
+    public PathParser(Location startLocation)
+    {
+        requireNonNull(startLocation, "startLocation is null");
+
+        int pathStartLine = startLocation.line;
+        int pathStartColumn = startLocation.column;
+        this.errorListener = new BaseErrorListener()
+        {
+            @Override
+            public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String message, RecognitionException e)
+            {
+                // re-position the exception relatively to the start of the path expression within the query
+                int lineInQuery = pathStartLine - 1 + line;
+                int columnInQuery = line == 1 ? pathStartColumn + 1 + charPositionInLine : charPositionInLine + 1;
+                throw new ParsingException(message, e, lineInQuery, columnInQuery);
+            }
+        };
+    }
+
+    public PathNode parseJsonPath(String path)
+    {
+        try {
+            // according to the SQL specification, the path language is case-sensitive in both identifiers and key words
+            JsonPathLexer lexer = new JsonPathLexer(CharStreams.fromString(path));
+            CommonTokenStream tokenStream = new CommonTokenStream(lexer);
+            JsonPathParser parser = new JsonPathParser(tokenStream);
+
+            parser.addParseListener(new PostProcessor(Arrays.asList(parser.getRuleNames()), parser));
+
+            lexer.removeErrorListeners();
```

Comment:
why do you need to remove? Are there any by default?

Reply:
I can't find out. I copied a pattern used in several places.